torchrl/envs/libs/vmas.py

```diff
@@ -170,37 +251,132 @@ def _build_env(
 
         return env
 
-    @set_gym_backend("gym")
+    def _get_default_group_map(self, agent_names: List[str]):
+        # This function performs the default grouping in vmas
+        # Agents with names "str_int" will be grouped in group name "str"
+        # If any of the agents to not follow the naming convention, we fall back
+        # to all agents in one group named "agents"
+        group_map = {}
+        follows_convention = True
+        for agent_name in agent_names:
+            # See if the agent follows the convention "name_int"
+            agent_name_split = agent_name.split("_")
+            if len(agent_name_split) == 1:
+                follows_convention = False
+            try:
+                int(agent_name_split[-1])
+            except ValueError:
+                follows_convention = False
+
+            if not follows_convention:
+                break
+
+            # Group it with other agents that follow the same convention
+            group_name = "_".join(agent_name_split[:-1])
+            if group_name in group_map:
+                group_map[group_name].append(agent_name)
+            else:
+                group_map[group_name] = [agent_name]
+
+        if not follows_convention:
+            group_map = MarlGroupMapType.ALL_IN_ONE_GROUP.get_group_map(agent_names)
+
+        # For BC-compatibility rename the "agent" group to "agents"
+        if "agent" in group_map:
+            agent_group = group_map["agent"]
+            group_map["agents"] = agent_group
+            del group_map["agent"]
+        return group_map
+
     def _make_specs(
         self, env: "vmas.simulator.environment.environment.Environment"  # noqa
     ) -> None:
-        # TODO heterogenous spaces
+        # Create and check group map
+        self.agent_names = [agent.name for agent in self.agents]
+        self.agent_names_to_indices_map = {
+            agent.name: i for i, agent in enumerate(self.agents)
+        }
+        if self.group_map is None:
+            self.group_map = self._get_default_group_map(self.agent_names)
+        elif isinstance(self.group_map, MarlGroupMapType):
+            self.group_map = self.group_map.get_group_map(self.agent_names)
+        check_marl_grouping(self.group_map, self.agent_names)
+
+        self.unbatched_action_spec = CompositeSpec(device=self.device)
+        self.unbatched_observation_spec = CompositeSpec(device=self.device)
+        self.unbatched_reward_spec = CompositeSpec(device=self.device)
+
+        self.het_specs = False
+        self.het_specs_map = {}
+        for group in self.group_map.keys():
+            (
+                group_observation_spec,
+                group_action_spec,
+                group_reward_spec,
+                group_info_spec,
+            ) = self._make_unbatched_group_specs(group)
+            self.unbatched_action_spec[group] = group_action_spec
```

Comment:
is `self.unbatched_action_spec` public on purpose?
Can we review the public attributes of VMAS? I feel there are a lot of them and most are undocumented. This will hamper development in the future. I would make anything that is not user facing private to give us more dev freedom.

Reply:
Sure I will make private all I don't need to be public.
This will be bc-breaking though as the methods will not be available anymore.

Regarding the unbatched_spec, these are needed when creating modules. Giving the specs with the batch_size to the modules will lead to errors #1018. These attirbutes are used in BenchMARL for example